processing/src/main/java/org/apache/druid/segment/nested/VariantColumnSerializer.java

```diff
@@ -299,155 +301,241 @@ public void serialize(ColumnValueSelector<? extends StructuredData> selector) th
-          ).add(rowCount);
+  public static class VariantColumnSerializerizer implements Serializer
+  {
+    private final String columnName;
+    private final ByteBuffer columnNameBytes;
+    private final Byte variantTypeSetByte;
+
+    private final DictionaryWriter<String> dictionaryWriter;
+    private final FixedIndexedWriter<Long> longDictionaryWriter;
+    private final FixedIndexedWriter<Double> doubleDictionaryWriter;
+    private final FrontCodedIntArrayIndexedWriter arrayDictionaryWriter;
+
+    private final SingleValueColumnarIntsSerializer encodedValueSerializer;
+    private final GenericIndexedWriter<ImmutableBitmap> bitmapIndexWriter;
+
+    private final FixedIndexedIntWriter arrayElementDictionaryWriter;
+    private final GenericIndexedWriter<ImmutableBitmap> arrayElementIndexWriter;
+    private final boolean writeDictionary;
+
+    private final DictionaryIdLookup dictionaryIdLookup;
+
+    public VariantColumnSerializerizer(
+        String columnName,
+        Byte variantTypeSetByte,
+        DictionaryWriter<String> dictionaryWriter,
+        FixedIndexedWriter<Long> longDictionaryWriter,
+        FixedIndexedWriter<Double> doubleDictionaryWriter,
+        FrontCodedIntArrayIndexedWriter arrayDictionaryWriter,
+        SingleValueColumnarIntsSerializer encodedValueSerializer,
+        GenericIndexedWriter<ImmutableBitmap> bitmapIndexWriter,
+        FixedIndexedIntWriter arrayElementDictionaryWriter,
+        GenericIndexedWriter<ImmutableBitmap> arrayElementIndexWriter,
+        DictionaryIdLookup dictionaryIdLookup,
+        boolean writeDictionary
+    )
+    {
+      this.columnName = columnName;
+      this.columnNameBytes = ColumnSerializerUtils.stringToUtf8InVSizeByteBuffer(columnName);
+      this.variantTypeSetByte = variantTypeSetByte;
+      this.dictionaryWriter = dictionaryWriter;
+      this.longDictionaryWriter = longDictionaryWriter;
+      this.doubleDictionaryWriter = doubleDictionaryWriter;
+      this.arrayDictionaryWriter = arrayDictionaryWriter;
+      this.encodedValueSerializer = encodedValueSerializer;
+      this.bitmapIndexWriter = bitmapIndexWriter;
+      this.arrayElementDictionaryWriter = arrayElementDictionaryWriter;
+      this.arrayElementIndexWriter = arrayElementIndexWriter;
+      this.writeDictionary = writeDictionary;
+      this.dictionaryIdLookup = dictionaryIdLookup;
+
+      boolean[] dictionariesSorted = new boolean[]{
+          dictionaryWriter.isSorted(),
+          longDictionaryWriter.isSorted(),
+          doubleDictionaryWriter.isSorted(),
+          arrayDictionaryWriter.isSorted()
+      };
+      for (boolean sorted : dictionariesSorted) {
+        if (writeDictionary && !sorted) {
+          throw DruidException.defensive(
+              "Dictionary is not sorted? [%s]  Should always be sorted",
+              Arrays.toString(dictionariesSorted)
+          );
         }
       }
```

Comment:
It prints `dictionariesSorted` which mentions which all dictionaries are not sorted. If that is too obscure, should I expand this into a separate check and error message for each dictionary?

Reply:
That would hide which all dictionaries are not sorted, since it's reduced to a single boolean. Here, the error message would look like `"Dictionary is not sorted? [[true, true, false, true, false]]  Should always be sorted"`, which does give the information on which dictionary the problem is with.